solarforecastarbiter/metrics/probabilistic.py

```diff
@@ -0,0 +1,210 @@
+    return BS
+
+
+def brier_skill_score(fx, fx_prob, ref, ref_prob, obs):
+    """Brier Skill Score (BSS).
+
+        BSS = 1 - BS_f / BS_ref
+
+    where BS_fx is the evaluated forecast and BS_ref is a reference forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    ref : (n,) array_like
+        Reference forecast (physical units).
+    ref_prob : (n,) array_like
+        Probability [%] associated with the reference forecast.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    skill : float
+        The Brier Skill Score [-].
+
+    """
+    bs_fx = brier_score(fx, fx_prob, obs)
+    bs_ref = brier_score(ref, ref_prob, obs)
+    return 1.0 - bs_fx / bs_ref
+
+
+def reliability(fx, fx_prob, obs):
+    """Reliability (REL) of the forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    REL : float
+        The reliability of the forecast, where a perfectly reliable forecast
+        has REL = 0.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities
+    if len(f) < 1000:
+        n_decimals = 1
+    else:
+        n_decimals = 2
+
+    f = np.around(f, decimals=n_decimals)
+    REL = 0.0
+    for f_i in np.unique(f):
+        N_i = len(f[f == f_i])  # no. forecasts per set
+        o_i = np.mean(o[f == f_i])    # mean event value per set
+        REL += N_i * (f_i - o_i) ** 2
+
+    REL = REL / len(f)
+    return REL
+
+
+def resolution(fx, fx_prob, obs):
+    """Resolution (RES) of the forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    RES : float
+        The resolution of the forecast, where higher values are better.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities
+    if len(f) < 1000:
+        n_decimals = 1
+    else:
+        n_decimals = 2
+
+    f = np.around(f, decimals=n_decimals)
+    RES = 0.0
+    o_avg = np.mean(o)
+    for f_i in np.unique(f):
+        N_i = len(f[f == f_i])  # no. forecasts per set
+        o_i = np.mean(o[f == f_i])    # mean event value per set
+        RES += N_i * (o_i - o_avg) ** 2
+
+    RES = RES / len(f)
+    return RES
```

Comment:
`reliability` and `resolution` calculations and thus functions are so similar. From an API standpoint, I think it's good to have both. For implementation, I wonder if we should write a single function `reliability_resolution` that returns both:

```python
def reliability_resolution(fx, fx_prob, obs):
    # calculation
    return rel, res

def reliability(fx, fx_probs, obs):
    return reliability_resolution(fx, fx_probs, obs)[0]

def resolution(fx, fx_probs, obs)
    return reliability_resolution(fx, fx_probs, obs)[1]
```

maybe a private function `_reliability_resolution` to keep the api simple? could add another public function to bundle all three?

```python
def reliability_resolution_uncertainty():
    return _reliability_resolution() + uncertainty()
```

🤷‍♂

Reply:
Along the same lines, I was wondering whether it would be "better" to have a `brier_decomposition()` function that returns the `reliability`, `resolution` and `uncertainty` (rather than separate functions). Alternatively, add a kwarg to `brier_score()` (e.g. `brier_score(decomposition=True)`) for when you want to also return the 3-component decomposition.

Reply:
`brier_decomposition` is a good name. I lean towards a separate function instead of a kwarg for `brier_score`.

Reply:
I'll add the `brier_decomposition()` function, which the `reliability()`, `resolution()` and `uncertainty()` functions will then call.

Reply:
Added the `brier_decomposition()` function, which the other functions call.